src/clr-angular/forms/combobox/option-items.directive.ts

```diff
@@ -0,0 +1,120 @@
+ */
+import { NgForOf, NgForOfContext } from '@angular/common';
+import {
+  Directive,
+  DoCheck,
+  Input,
+  IterableDiffer,
+  IterableDiffers,
+  TemplateRef,
+  TrackByFunction,
+  ViewContainerRef,
+  OnDestroy,
+} from '@angular/core';
+
+import { OptionSelectionService } from './providers/option-selection.service';
+import { Subscription } from 'rxjs';
+import { ClrPopoverPositionService } from '../../utils/popover/providers/popover-position.service';
+
+@Directive({
+  selector: '[clrOptionItems][clrOptionItemsOf]',
+})
+export class ClrOptionItems<T> implements DoCheck, OnDestroy {
+  private iterableProxy: NgForOf<T>;
+  private _rawItems: T[];
+  private filteredItems: T[];
+  private subscriptions: Subscription[] = [];
+  private filter = '';
+  private _filterField: string;
+  private differ: IterableDiffer<T> | null = null;
+
+  @Input('clrOptionItemsOf')
+  public set rawItems(items: T[]) {
+    this._rawItems = items ? items : [];
+    this.updateItems();
+  }
+
+  @Input('clrOptionItemsTrackBy')
+  set trackBy(value: TrackByFunction<T>) {
+    this.iterableProxy.ngForTrackBy = value;
+  }
+
+  @Input('clrOptionItemsField')
+  public set field(field: string) {
+    this._filterField = field;
+    this.optionService.displayField = field;
+  }
+
+  constructor(
+    public template: TemplateRef<NgForOfContext<T>>,
+    private differs: IterableDiffers,
+    private optionService: OptionSelectionService<T>,
+    private positionService: ClrPopoverPositionService,
+    private vcr: ViewContainerRef
+  ) {
+    this.iterableProxy = new NgForOf<T>(this.vcr, this.template, this.differs);
+    this.subscriptions.push(
+      optionService.inputChanged.subscribe(filter => {
+        this.filter = filter;
+        this.updateItems();
+      })
+    );
+  }
+
+  private updateItems() {
+    if (!this._rawItems || this.filter === undefined || this.filter === null) {
+      return;
+    }
+    if (this._filterField) {
+      this.filteredItems = this._rawItems.filter(item => {
+        const objValue = (item as any)[this._filterField];
+        return objValue !== undefined
+          ? String(objValue)
+              .toLowerCase()
+              .indexOf(String(this.filter.toLowerCase())) > -1
+          : false;
+      });
+    } else {
+      // Filter by all item object values
+      this.filteredItems = this._rawItems.filter(item => {
+        if (typeof item !== 'object') {
+          return (
+            String(item)
+              .toLowerCase()
+              .indexOf(String(this.filter).toLowerCase()) > -1
+          );
+        }
+        const objValues = Object.values(item).filter(value => {
+          return (
+            String(value)
+              .toLowerCase()
+              .indexOf(String(this.filter).toLowerCase()) > -1
+          );
+        });
+        return objValues.length > 0;
+      });
+    }
+    this.iterableProxy.ngForOf = this.filteredItems;
+  }
+
+  ngDoCheck() {
+    this.iterableProxy.ngDoCheck();
```

Comment:
Is there a reason for calling the iterableProxy.ngDoCheck here instead of on line #101?IIR - the iterableProxy needs the component to manually call ngDoCheck whenever items are changed and that propagates and allows the app to detecct changes. 

I believe when it is called in `updateItems` the ngDoChecks will be picked up here as well and run. But perhaps there is a perfermance issue I'm not seeing.

Reply:
If I move it to 101 it seems to work, but it does start to fail test cases in the 'correctly mutates and resets an array with trackBy' test.
Also, it seems normal that when you wrap some component, and you redefine some of their lifecycle methods, you should call the basic component's methods from your redefined methods. Otherwise you're messing with the natural lifecycle of the original component.
If you have a feeling that the other approach is better, please try it, check it against the tests and try to prove that it really is better.
I am open for f2f discussion, won't be closing this conversation.

Reply:
We looked at the code together and found an even better solution - keep the call still in the local ngDoCheck, but inside the block that tests for differences. 
It works with the same efficiency as the suggested by Matt call-from-updateItems scenario, but does not break the trackBy tests.